File: src/main/java/org/cubexmc/railway/service/virtual/VirtualTrainPool.java
package org.cubexmc.railway.service.virtual;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.PriorityQueue;
import java.util.Set;
import java.util.UUID;
import java.util.logging.Logger;

import org.cubexmc.railway.service.virtual.VirtualTrain.EventType;

import org.cubexmc.railway.estimation.TravelTimeEstimator;
import org.cubexmc.railway.model.Line;

/**
 * Manages a pool of virtual trains for a single line.
 * Virtual trains run continuously in the background and can be materialized
 * when players are nearby.
 */
public class VirtualTrainPool {

    private final String lineId;
    private final Map<UUID, VirtualTrain> virtualTrains = new HashMap<>(); // Replaced List with Map for faster lookups
    private final PriorityQueue<TrainEvent> eventQueue;
    private final Set<UUID> materializedIds = new HashSet<>();
    private final int dwellTicks;

    // Internal event class for the PriorityQueue
    private static class TrainEvent {
        final long tick;
        final UUID trainId;
        final EventType type;

        TrainEvent(long tick, UUID trainId, EventType type) {
            this.tick = tick;
            this.trainId = trainId;
            this.type = type;
        }
    }

    /**
     * Create a new virtual train pool for a line.
     * 
     * @param lineId     The line ID
     * @param dwellTicks Dwell time at each stop in ticks
     */
    public VirtualTrainPool(String lineId, int dwellTicks) {
        this.lineId = lineId;
        this.dwellTicks = Math.max(20, dwellTicks);
        this.eventQueue = new PriorityQueue<>(Comparator.comparingLong(e -> e.tick));
    }

    /**
     * Initialize the pool with virtual trains based on headway.
     * Trains are distributed evenly along the line.
     * 
     * @param line           The line configuration
     * @param headwaySeconds Headway between trains in seconds
     * @param estimator      Travel time estimator
     * @param currentTick    Current server tick
     */
    public void initialize(Line line, int headwaySeconds, TravelTimeEstimator estimator, long currentTick) {
        virtualTrains.clear();
        eventQueue.clear();
        materializedIds.clear();

        List<String> stopIds = line.getOrderedStopIds();
        if (stopIds == null || stopIds.size() < 2) {
            return;
        }

        // Calculate total cycle time (travel + dwell at each stop)
        double totalTravelSeconds = 0;
        for (int i = 0; i < stopIds.size() - 1; i++) {
            totalTravelSeconds += estimator.estimateSeconds(lineId, stopIds.get(i), stopIds.get(i + 1));
        }
        double dwellSecondsPerStop = dwellTicks / 20.0;
        double totalCycleSeconds = totalTravelSeconds + (stopIds.size() * dwellSecondsPerStop);

        if (totalCycleSeconds <= 0) {
            totalCycleSeconds = stopIds.size() * 30.0; // Fallback
        }

        // Safety check for estimator to prevent NaN/Infinity issues
        if (Double.isNaN(totalCycleSeconds) || Double.isInfinite(totalCycleSeconds)) {
            totalCycleSeconds = stopIds.size() * 30.0;
        }

        // Determine number of virtual trains
        int trainCount = Math.max(1, (int) Math.ceil(totalCycleSeconds / Math.max(10, headwaySeconds)));

        // Cap at reasonable number
        trainCount = Math.min(trainCount, 10);

        // Calculate segment boundaries for distribution
        List<SegmentBoundary> boundaries = buildSegmentBoundaries(stopIds, estimator);
        double totalTime = boundaries.isEmpty() ? totalCycleSeconds
                : boundaries.get(boundaries.size() - 1).cumulativeTime;

        // Distribute trains evenly
        for (int i = 0; i < trainCount; i++) {
            double timeOffset = (totalTime / trainCount) * i;

            // Find segment and progress for this time offset
            int stopIndex = 0;
            double progress = 0.0;

            for (int j = 0; j < boundaries.size(); j++) {
                SegmentBoundary seg = boundaries.get(j);
                if (timeOffset <= seg.cumulativeTime) {
                    stopIndex = seg.fromStopIndex;
                    double segmentStart = j > 0 ? boundaries.get(j - 1).cumulativeTime : 0;
                    double segmentDuration = seg.cumulativeTime - segmentStart;
                    if (segmentDuration > 0) {
                        double withinSegment = timeOffset - segmentStart;
                        if (withinSegment <= dwellSecondsPerStop) {
                            // Still in dwell time
                            progress = 0.0;
                        } else {
                            // In transit
                            double travelTime = segmentDuration - dwellSecondsPerStop;
                            if (travelTime > 0) {
                                progress = Math.min(1.0, (withinSegment - dwellSecondsPerStop) / travelTime);
                            }
                        }
                    }
                    break;
                }
                stopIndex = seg.toStopIndex;
            }

            VirtualTrain vt = new VirtualTrain(lineId, stopIds, dwellTicks, stopIndex, progress, currentTick);
            virtualTrains.put(vt.getId(), vt);

            // Schedule first event
            scheduleNextEvent(vt, estimator, currentTick);
        }
    }

    private List<SegmentBoundary> buildSegmentBoundaries(List<String> stopIds, TravelTimeEstimator estimator) {
        List<SegmentBoundary> boundaries = new ArrayList<>();
        double cumulative = 0;
        double dwellSeconds = dwellTicks / 20.0;

        for (int i = 0; i < stopIds.size() - 1; i++) {
            cumulative += dwellSeconds; // Dwell at departure stop
            cumulative += estimator.estimateSeconds(lineId, stopIds.get(i), stopIds.get(i + 1));
            boundaries.add(new SegmentBoundary(i, i + 1, cumulative));
        }
        return boundaries;
    }

    private static class SegmentBoundary {
        final int fromStopIndex;
        final int toStopIndex;
        final double cumulativeTime;

        SegmentBoundary(int from, int to, double time) {
            this.fromStopIndex = from;
            this.toStopIndex = to;
            this.cumulativeTime = time;
        }
    }

    /**
     * Update all virtual trains.
     * 
     * @param currentTick Current server tick
     * @param estimator   Travel time estimator
     */
    /**
     * Update all virtual trains by processing events up to currentTick.
     * This is the core Discrete Event Simulation loop.
     * 
     * @param currentTick Current server tick
     * @param estimator   Travel time estimator
     */
    public void tick(long currentTick, TravelTimeEstimator estimator) {
        // Process all events that are due (scheduled time <= current time)
        while (!eventQueue.isEmpty() && eventQueue.peek().tick <= currentTick) {
            TrainEvent event = eventQueue.poll();

            VirtualTrain vt = virtualTrains.get(event.trainId);
            if (vt == null)
                continue; // Train removed?

            // If materialized, we still process the event conceptually to keep the virtual
            // state in sync?
            // Or do we pause?
            // The polling version said: "Skip materialized trains".
            // If we skip processing, the train effectively pauses in virtual space.
            // When it un-materializes, it calls returnToVirtual which resets state.
            // SO: If materialized, we should probably NOT update the virtual state
            // automatically
            // because the physical train is now the authority.
            // WE JUST DROP THE EVENT. The returnToVirtual will re-schedule events.

            if (materializedIds.contains(vt.getId())) {
                continue;
            }

            // Validate event order (sanity check)
            if (vt.getNextEventTick() != event.tick) {
                // Old/stale event? In a robust system we'd handle this.
                // For now, trust the queue but verifies against train's state
                // If the train was somehow updated otherwise, this event might be stale.
            }

            // Process the event
            // 1. Update Train State
            // We need to calculate the *result* of this event.
            // If Arrival -> State becomes Waiting, next event is Departure
            // If Departure -> State becomes Moving, next event is Arrival at next stop

            // CALCULATE DATA FOR NEXT STEP
            int currentStop = vt.getCurrentStopIndex();

            if (event.type == EventType.ARRIVAL) {
                // Train just arrived at 'currentStop'
                // It stays there for dwellTicks
                // Update train state to: WAITING at currentStop
                // NOTE: 'onEvent' takes the state *after* the event.
                // ARRIVAL -> sets type=ARRIVAL (Waiting)
                vt.onEvent(EventType.ARRIVAL, currentStop, event.tick, 0); // duration irrelevant for waiting

                // Schedule next: DEPARTURE
                eventQueue.add(new TrainEvent(vt.getNextEventTick(), vt.getId(), EventType.DEPARTURE));

            } else if (event.type == EventType.DEPARTURE) {
                // Train is departing from 'currentStop'
                // It will go to 'currentStop + 1'
                // Calculate duration
                List<String> stopIds = estimator.getPlugin().getLineManager().getLine(lineId).getOrderedStopIds();
                int nextIndex = currentStop + 1;

                // Handle Loop/Terminal logic
                if (nextIndex >= stopIds.size()) {
                    // Terminal reached
                    boolean isLoop = serviceIsLoop(stopIds);
                    if (isLoop) {
                        nextIndex = 1; // 0 is start/end duplicate usually?
                        // Actually let's assume standard loop: A->B->C->A
                        // If stops are [A, B, C, A], size is 4. indices 0,1,2,3.
                        // Arrive at 3 (A). Depart from 3?
                        // Usually loop lines just run continuously.
                        // Let's use 1 if indices match start/end?
                        // Or maybe index 0?
                        // If stopIds[0].equals(stopIds[last]), then arrival at last == arrival at
                        // first.
                        // The previous logic in VirtualTrain handle terminal:
                        // "currentStopIndex = 0; target = -1; state = WAITING"
                        // So it basically teleports to start and waits.

                        // Let's simulate that:
                        // Departure from Terminal -> Instant teleport to Start ARRIVAL?
                        // Or just set next index to 1?
                        // Let's rely on TravelTimeEstimator to give us time from Last -> First?
                        // If the line defines A->B->...->A, then Last->First is dist 0 ?
                        // The estimator likely returns 0 or small value.

                        // SIMPLIFICATION:
                        // If at terminal, we simply map back to 0 immediately.
                        nextIndex = 0;
                        // And we treat this as a "reset" event.
                        // But wait, the standard logic was: "wait at start".
                        // So Departure from Terminal -> Arrival at Start (Duration 0/Small)
                    } else {
                        // Non-loop. Turnaround.
                        // "Departure" from terminal -> Arrival at start (Teleport)
                        nextIndex = 0;
                    }
                }

                // Calculate travel time
                String fromId = stopIds.get(currentStop);
                // Be careful if nextIndex was wrapped
                if (nextIndex >= stopIds.size())
                    nextIndex = 0; // Safety for non-loop
                String toId = stopIds.get(nextIndex);

                double duration = estimator.estimateSeconds(lineId, fromId, toId);
                if (duration <= 0)
                    duration = 10.0;

                // Update train state
                vt.onEvent(EventType.DEPARTURE, currentStop, event.tick, duration);
                // Accessing package-private setTargetStopIndex if needed, but onEvent sets it.
                // However, internal logic in onEvent sets target = stop + 1.
                // We might need to correct it if we wrapped around.
                vt.setTargetStopIndex(nextIndex);

                // Schedule next: ARRIVAL
                eventQueue.add(new TrainEvent(vt.getNextEventTick(), vt.getId(), EventType.ARRIVAL));
            }
        }
    }

    private boolean serviceIsLoop(List<String> stopIds) {
        return stopIds.size() >= 2 && stopIds.get(0).equals(stopIds.get(stopIds.size() - 1));
    }

    private void scheduleNextEvent(VirtualTrain vt, TravelTimeEstimator estimator, long currentTick) {
        // Should calculate based on current state where the next event is
        // Used during initialization or restoration

        // If last was ARRIVAL -> Next is DEPARTURE
        if (vt.getLastEventType() == EventType.ARRIVAL) {
            long nextTick = vt.getLastEventTick() + dwellTicks;
            // Ensure it's in future? If past, schedule for now?
            // For initialization it might be in past if we used currentTick as base.
            // But we want to distribute them.

            // If we initialized a train "mid-dwell", nextTick might be soon.
            vt.setNextEventTick(nextTick);
            eventQueue.add(new TrainEvent(nextTick, vt.getId(), EventType.DEPARTURE));
        } else {
            // If last was DEPARTURE -> Next is ARRIVAL
            // need duration
            // We can't easily get duration here without the logic from tick().
            // But valid state implies currentPathDurationSeconds is set.
            // So we just use getNextEventTick() which was set in constructor/restore
            eventQueue.add(new TrainEvent(vt.getNextEventTick(), vt.getId(), EventType.ARRIVAL));
        }
    }

    /**
     * Find the best virtual train candidate to materialize for a given stop.
     * Returns the train with the smallest ETA that is not already materialized.
     * 
     * @param stopId        Target stop ID
     * @param stopIds       Ordered stop IDs for the line
     * @param maxEtaSeconds Maximum ETA in seconds
     * @param estimator     Travel time estimator
     * @return Best candidate, or null if none found
     */
    public VirtualTrain findBestCandidateForStop(String stopId, List<String> stopIds,
            double maxEtaSeconds, TravelTimeEstimator estimator) {
        int stopIndex = stopIds.indexOf(stopId);
        if (stopIndex < 0)
            return null;

        VirtualTrain best = null;
        double bestEta = Double.POSITIVE_INFINITY;

        for (VirtualTrain vt : virtualTrains.values()) {
            if (materializedIds.contains(vt.getId())) {
                continue;
            }
            if (vt.isAtTerminal()) {
                continue;
            }

            double eta = vt.estimateEtaToStop(stopIndex, estimator);
            if (Double.isFinite(eta) && eta <= maxEtaSeconds && eta < bestEta) {
                bestEta = eta;
                best = vt;
            }
        }

        return best;
    }

    /**
     * Mark a virtual train as materialized (physical instance created).
     */
    public void markMaterialized(UUID virtualTrainId) {
        materializedIds.add(virtualTrainId);
    }

    /**
     * Check if a virtual train is currently materialized.
     */
    public boolean isMaterialized(UUID virtualTrainId) {
        return materializedIds.contains(virtualTrainId);
    }

    /**
     * Clear materialized flag without restoring state.
     * Used when the physical train was removed but the virtual train
     * should continue from its current simulated position.
     */
    public void clearMaterialized(UUID virtualTrainId) {
        materializedIds.remove(virtualTrainId);
    }

    /**
     * Return a physical train back to virtual state.
     * 
     * @param virtualTrainId The virtual train ID
     * @param stopIndex      Current stop index
     * @param targetIndex    Target stop index (-1 if waiting)
     * @param progress       Segment progress (0-1)
     * @param isWaiting      Whether the train was waiting
     * @param currentTick    Current tick
     */
    public void returnToVirtual(UUID virtualTrainId, int stopIndex, int targetIndex,
            double progress, boolean isWaiting, long currentTick) {
        materializedIds.remove(virtualTrainId);

        for (VirtualTrain vt : virtualTrains.values()) {
            if (vt.getId().equals(virtualTrainId)) {
                vt.restoreState(stopIndex, targetIndex, progress, isWaiting, currentTick);
                // Re-schedule events since it's back in virtual control
                // We need to fetch estimator... ugly dependency issue here as estimator not
                // passed to returnToVirtual
                // We will defer scheduling to the next tick() call?
                // No, tick() only processes existing events. If queue is empty for this train,
                // it never moves.

                // We MUST schedule the next event here.
                // But we don't have the estimator.
                // Solution: We'll rely on the restored state's 'nextEventTick' (which is set to
                // current + buffer)
                // And just add that to queue.
                if (isWaiting) {
                    // It was waiting, so next is DEPARTURE
                    eventQueue.add(new TrainEvent(vt.getNextEventTick(), vt.getId(), EventType.DEPARTURE));
                } else {
                    // It was moving, so next is ARRIVAL
                    eventQueue.add(new TrainEvent(vt.getNextEventTick(), vt.getId(), EventType.ARRIVAL));
                }

                break;
            }
        }
    }

    /**
     * Completely remove a virtual train (e.g., when it completes a non-loop line).
     * A new one will be spawned at the start on the next initialize call.
     */
    public void removeVirtualTrain(UUID virtualTrainId) {
        materializedIds.remove(virtualTrainId);
        virtualTrains.remove(virtualTrainId);
        // We leave the events in the queue; they will be ignored in tick() because
        // virtualTrains.get(id) will return null.
    }

    /**
     * Get all virtual trains (for debugging/monitoring).
     */
    public List<VirtualTrain> getVirtualTrains() {
        return new ArrayList<>(virtualTrains.values());
    }

    /**
     * Get count of active (non-terminal) virtual trains.
     */
    public int getActiveCount() {
        int count = 0;
        for (VirtualTrain vt : virtualTrains.values()) {
            if (!vt.isAtTerminal()) {
                count++;
            }
        }
        return count;
    }

    /**
     * Check if there are any available (non-materialized, non-terminal) virtual
     * trains.
     */
    public boolean hasAvailableTrains() {
        for (VirtualTrain vt : virtualTrains.values()) {
            if (!materializedIds.contains(vt.getId()) && !vt.isAtTerminal()) {
                return true;
            }
        }
        return false;
    }

    public String getLineId() {
        return lineId;
    }

    @Override
    public String toString() {
        return String.format("VirtualTrainPool[line=%s, trains=%d, materialized=%d]",
                lineId, virtualTrains.size(), materializedIds.size());
    }
}

File: src/main/java/org/cubexmc/metro/service/virtual/VirtualTrainPool.kt
package org.cubexmc.metro.service.virtual

import java.util.PriorityQueue
import java.util.UUID
import kotlin.math.ceil
import kotlin.math.max
import kotlin.math.min
import org.cubexmc.metro.estimation.TravelTimeEstimator
import org.cubexmc.metro.model.Line
import org.cubexmc.metro.service.virtual.VirtualTrain.EventType
import org.cubexmc.metro.util.LineTopologyUtil

/**
 * Discrete-event pool of virtual trains for one line.
 */
class VirtualTrainPool(
    val lineId: String?,
    dwellTicks: Int,
) {
    private val virtualTrains: MutableMap<UUID, VirtualTrain> = HashMap()
    private val eventQueue: PriorityQueue<TrainEvent> =
        PriorityQueue(Comparator.comparingLong { event: TrainEvent -> event.tick })
    private val materializedIds: MutableSet<UUID> = HashSet()
    private val dwellTicks: Int = max(20, dwellTicks)
    private var cachedStopIds: List<String> = ArrayList()

    private class TrainEvent(
        val tick: Long,
        val trainId: UUID,
        val type: EventType,
    )

    private class SegmentBoundary(
        val fromStopIndex: Int,
        val toStopIndex: Int,
        val cumulativeTime: Double,
    )

    fun initialize(line: Line, headwaySeconds: Int, estimator: TravelTimeEstimator, currentTick: Long) {
        virtualTrains.clear()
        eventQueue.clear()
        materializedIds.clear()

        val stopIds = orderedStopIdsOrNull(line)
        if (stopIds == null || stopIds.size < 2) {
            return
        }
        cachedStopIds = ArrayList(stopIds)

        var totalTravelSeconds = 0.0
        for (index in 0 until stopIds.size - 1) {
            totalTravelSeconds += estimator.estimateSeconds(lineId, stopIds[index], stopIds[index + 1])
        }
        val dwellSecondsPerStop = dwellTicks / 20.0
        var totalCycleSeconds = totalTravelSeconds + stopIds.size * dwellSecondsPerStop
        if (totalCycleSeconds <= 0) {
            totalCycleSeconds = stopIds.size * 30.0
        }
        if (totalCycleSeconds.isNaN() || totalCycleSeconds.isInfinite()) {
            totalCycleSeconds = stopIds.size * 30.0
        }

        var trainCount = max(1, ceil(totalCycleSeconds / max(10, headwaySeconds)).toInt())
        trainCount = min(trainCount, 10)

        val boundaries = buildSegmentBoundaries(stopIds, estimator)
        val totalTime = if (boundaries.isEmpty()) {
            totalCycleSeconds
        } else {
            boundaries[boundaries.size - 1].cumulativeTime
        }

        for (index in 0 until trainCount) {
            val timeOffset = totalTime / trainCount * index
            var stopIndex = 0
            var targetIndex = min(1, stopIds.size - 1)
            var progress = 0.0
            var waiting = true
            var elapsedDwellSeconds = 0.0
            var segmentDurationSeconds = estimator.estimateSeconds(lineId, stopIds[0], stopIds[targetIndex])

            for (boundaryIndex in boundaries.indices) {
                val segment = boundaries[boundaryIndex]
                if (timeOffset <= segment.cumulativeTime) {
                    stopIndex = segment.fromStopIndex
                    targetIndex = segment.toStopIndex
                    val segmentStart = if (boundaryIndex > 0) {
                        boundaries[boundaryIndex - 1].cumulativeTime
                    } else {
                        0.0
                    }
                    val segmentDuration = segment.cumulativeTime - segmentStart
                    segmentDurationSeconds = max(0.05, segmentDuration - dwellSecondsPerStop)
                    if (segmentDuration > 0) {
                        val withinSegment = timeOffset - segmentStart
                        if (withinSegment <= dwellSecondsPerStop) {
                            waiting = true
                            elapsedDwellSeconds = withinSegment
                            progress = 0.0
                        } else {
                            waiting = false
                            if (segmentDurationSeconds > 0) {
                                progress = min(
                                    1.0,
                                    (withinSegment - dwellSecondsPerStop) / segmentDurationSeconds,
                                )
                            }
                        }
                    }
                    break
                }
                stopIndex = segment.toStopIndex
                targetIndex = min(stopIndex + 1, stopIds.size - 1)
            }

            val train = VirtualTrain(lineId, stopIds, dwellTicks, stopIndex, progress, currentTick)
            if (waiting) {
                train.initializeWaitingAt(stopIndex, elapsedDwellSeconds, currentTick)
            } else {
                train.initializeMovingBetween(stopIndex, targetIndex, progress, segmentDurationSeconds, currentTick)
            }
            virtualTrains[train.id] = train
            scheduleNextEvent(train, estimator, currentTick)
        }
    }

    /** Line is Kotlin now, but this preserves the old Java null guard for mocks. */
    private fun orderedStopIdsOrNull(line: Line): List<String>? = line.orderedStopIds

    private fun buildSegmentBoundaries(
        stopIds: List<String>,
        estimator: TravelTimeEstimator,
    ): List<SegmentBoundary> {
        val boundaries = ArrayList<SegmentBoundary>()
        var cumulative = 0.0
        val dwellSeconds = dwellTicks / 20.0
        for (index in 0 until stopIds.size - 1) {
            cumulative += dwellSeconds
            cumulative += estimator.estimateSeconds(lineId, stopIds[index], stopIds[index + 1])
            boundaries.add(SegmentBoundary(index, index + 1, cumulative))
        }
        return boundaries
    }

    fun tick(currentTick: Long, estimator: TravelTimeEstimator) {
        while (eventQueue.isNotEmpty() && eventQueue.peek().tick <= currentTick) {
            val event = eventQueue.poll()
            val train = virtualTrains[event.trainId] ?: continue
            if (materializedIds.contains(train.id)) {
                continue
            }

            val currentStop = train.currentStopIndex
            if (event.type == EventType.ARRIVAL) {
                train.onEvent(EventType.ARRIVAL, currentStop, event.tick, 0.0, cachedStopIds)
                eventQueue.add(TrainEvent(train.nextEventTick, train.id, EventType.DEPARTURE))
                continue
            }

            val stopIds = cachedStopIds
            var nextIndex = currentStop + 1
            if (nextIndex >= stopIds.size) {
                if (serviceIsLoop(stopIds)) {
                    nextIndex = 1
                    nextIndex = 0
                } else {
                    nextIndex = 0
                }
            }

            val fromId = stopIds[currentStop]
            if (nextIndex >= stopIds.size) {
                nextIndex = 0
            }
            val toId = stopIds[nextIndex]
            var duration = estimator.estimateSeconds(lineId, fromId, toId)
            if (duration <= 0) {
                duration = 10.0
            }

            train.onEvent(EventType.DEPARTURE, currentStop, event.tick, duration, stopIds)
            train.targetStopIndex = nextIndex
            eventQueue.add(TrainEvent(train.nextEventTick, train.id, EventType.ARRIVAL))
        }
    }

    private fun serviceIsLoop(stopIds: List<String>): Boolean = LineTopologyUtil.isLoop(stopIds)

    private fun scheduleNextEvent(
        train: VirtualTrain,
        estimator: TravelTimeEstimator,
        currentTick: Long,
    ) {
        if (train.lastEventType == EventType.ARRIVAL) {
            val nextTick = train.lastEventTick + dwellTicks
            train.nextEventTick = nextTick
            eventQueue.add(TrainEvent(nextTick, train.id, EventType.DEPARTURE))
        } else {
            eventQueue.add(TrainEvent(train.nextEventTick, train.id, EventType.ARRIVAL))
        }
    }

    fun findBestCandidateForStop(
        stopId: String,
        stopIds: List<String>,
        maxEtaSeconds: Double,
        estimator: TravelTimeEstimator,
        currentTick: Long,
    ): VirtualTrain? {
        val stopIndex = stopIds.indexOf(stopId)
        if (stopIndex < 0) {
            return null
        }

        var best: VirtualTrain? = null
        var bestEta = Double.POSITIVE_INFINITY
        for (train in virtualTrains.values) {
            if (materializedIds.contains(train.id) || train.isAtTerminal(stopIds)) {
                continue
            }
            val eta = train.estimateEtaToStop(stopIndex, estimator, currentTick)
            if (eta.isFinite() && eta <= maxEtaSeconds && eta < bestEta) {
                bestEta = eta
                best = train
            }
        }
        return best
    }

    fun markMaterialized(virtualTrainId: UUID) {
        materializedIds.add(virtualTrainId)
    }

    fun isMaterialized(virtualTrainId: UUID): Boolean = materializedIds.contains(virtualTrainId)

    fun clearMaterialized(virtualTrainId: UUID) {
        materializedIds.remove(virtualTrainId)
    }

    fun releaseMaterialized(virtualTrainId: UUID) {
        if (!materializedIds.remove(virtualTrainId)) {
            return
        }
        val train = virtualTrains[virtualTrainId] ?: return
        val nextType = if (train.lastEventType == EventType.ARRIVAL) {
            EventType.DEPARTURE
        } else {
            EventType.ARRIVAL
        }
        eventQueue.add(TrainEvent(train.nextEventTick, train.id, nextType))
    }

    fun returnToVirtual(
        virtualTrainId: UUID,
        stopIndex: Int,
        targetIndex: Int,
        progress: Double,
        isWaiting: Boolean,
        currentTick: Long,
        stopIds: List<String>,
    ) {
        materializedIds.remove(virtualTrainId)
        for (train in virtualTrains.values) {
            if (train.id == virtualTrainId) {
                train.restoreState(stopIndex, targetIndex, progress, isWaiting, currentTick, stopIds)
                val nextType = if (isWaiting) EventType.DEPARTURE else EventType.ARRIVAL
                eventQueue.add(TrainEvent(train.nextEventTick, train.id, nextType))
                break
            }
        }
    }

    fun removeVirtualTrain(virtualTrainId: UUID) {
        materializedIds.remove(virtualTrainId)
        virtualTrains.remove(virtualTrainId)
    }

    fun getVirtualTrains(): List<VirtualTrain> = ArrayList(virtualTrains.values)

    fun getActiveCount(stopIds: List<String>): Int {
        var count = 0
        for (train in virtualTrains.values) {
            if (!train.isAtTerminal(stopIds)) {
                count++
            }
        }
        return count
    }

    fun hasAvailableTrains(stopIds: List<String>): Boolean {
        for (train in virtualTrains.values) {
            if (!materializedIds.contains(train.id) && !train.isAtTerminal(stopIds)) {
                return true
            }
        }
        return false
    }

    fun refreshTopology(newStopIds: List<String>, estimator: TravelTimeEstimator, currentTick: Long) {
        val oldStopIds = ArrayList(cachedStopIds)
        cachedStopIds = ArrayList(newStopIds)
        eventQueue.clear()
        for (train in virtualTrains.values) {
            train.syncToNewTopology(oldStopIds, newStopIds, estimator, currentTick)
            if (materializedIds.contains(train.id)) {
                continue
            }
            scheduleNextEvent(train, estimator, currentTick)
        }
    }

    override fun toString(): String = String.format(
        "VirtualTrainPool[line=%s, trains=%d, materialized=%d]",
        lineId,
        virtualTrains.size,
        materializedIds.size,
    )
}
